ui/lib/components/AutoRefreshButton/AutoRefreshButton.tsx

```diff
@@ -0,0 +1,203 @@
+  onAutoRefreshSecondsChange: (v: number) => void
+  onRefresh: () => void
+  // set to false will pause the auto refresh
+  disabled?: boolean
+}
+
+const translations = {
+  en: {
+    refresh: 'Refresh',
+    auto_refresh: {
+      title: 'Auto Refresh',
+      off: 'Off',
+    },
+  },
+  zh: {
+    refresh: '刷新',
+    auto_refresh: {
+      title: '自动刷新',
+      off: '关闭',
+    },
+  },
+}
+
+for (const key in translations) {
+  addTranslationResource(key, {
+    component: {
+      autoRefreshButton: translations[key],
+    },
+  })
+}
+
+export function AutoRefreshButton({
+  autoRefreshSecondsOptions,
+  autoRefreshSeconds,
+  onAutoRefreshSecondsChange,
+  onRefresh,
+  disabled = false,
+}: AutoRefreshButtonProps) {
+  const { t } = useTranslation()
+  const autoRefreshMenu = useMemo(
+    () => (
+      <Menu
+        onClick={({ key }) =>
+          onAutoRefreshSecondsChange(parseInt(key as string))
+        }
+        selectedKeys={[String(autoRefreshSeconds || 0)]}
+      >
+        <Menu.ItemGroup
+          title={t('component.autoRefreshButton.auto_refresh.title')}
+        >
+          <Menu.Item key="0">
+            {t('component.autoRefreshButton.auto_refresh.off')}
+          </Menu.Item>
+          <Menu.Divider />
+          {autoRefreshSecondsOptions.map((sec) => {
+            return (
+              <Menu.Item key={String(sec)}>
+                {getValueFormat('s')(sec, 0)}
+              </Menu.Item>
+            )
+          })}
+        </Menu.ItemGroup>
+      </Menu>
+    ),
+    [autoRefreshSeconds, autoRefreshSecondsOptions, onAutoRefreshSecondsChange]
+  )
+
+  const timer = useRef<ReturnType<typeof setTimeout> | undefined>(undefined)
+  const [getRemainingRefreshSeconds, setRemainingRefreshSeconds] =
+    useGetSet(autoRefreshSeconds)
+
+  const resetTimer = useCallback(() => {
+    clearTimeout(timer.current!)
+    timer.current = undefined
+    setRemainingRefreshSeconds(autoRefreshSeconds)
+  }, [autoRefreshSeconds])
+
+  useEffect(() => {
+    clearTimeout(timer.current!)
+    timer.current = undefined
+    if (
+      // If remaining seconds is less than the new autoRefreshSeconds, keep the current remaining seconds.
+      // Otherwise, set remaining seconds to new autoRefreshSeconds.
+      getRemainingRefreshSeconds() > autoRefreshSeconds ||
+      getRemainingRefreshSeconds() === 0
+    ) {
+      setRemainingRefreshSeconds(autoRefreshSeconds)
+    }
+  }, [autoRefreshSeconds])
+
+  const handleRefresh = useCallback(async () => {
+    if (disabled) {
+      return
+    }
+    resetTimer()
+    onRefresh()
+  }, [disabled, resetTimer, onRefresh])
+
+  useEffect(() => {
+    // stop or pause auto refresh need to clear timer
+    if (!autoRefreshSeconds || disabled) {
+      if (!!timer.current) {
+        clearTimeout(timer.current)
+        timer.current = undefined
+      }
+      return
+    }
+
+    timer.current = setTimeout(() => {
+      if (getRemainingRefreshSeconds() === 0) {
```

Comment:
I guess you may like to have something like `onRemainingSecondsChange` to let the outside to control your remaining seconds. Otherwise it is not possible for the outside th reset the remaining seconds.

Reply:
Very helpful! It's looks like the concept of `remaining seconds` is unavoidable.